### PR TITLE
[pigeon] implemented generics for fields

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/63468

~~depends on: https://github.com/flutter/packages/pull/414~~

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:89]: This is pretty hard to parse; maybe adding an example " (e.g., `String` for `List<String>`)" would be helpful.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:97]: Wouldn't it be more useful to turn this into the type string people are used to dealing with? I.e., so that it would print:
```
(TypeArgument: List<String?>)
```

rather than:

```
(TypeArgument dataType:List isNullable:false typeArguments:[(TypeArgument dataType:String isNullable:true typeArguments:null)])
```

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:188]: Why don't these lines need to output all of the type information? If the type is "List<String>?", doesn't it need to print that instead of "List"?

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:267]: A comment with an example of what exactly the flattened representation is would be useful.

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:275]: A declaration comment would be useful here. E.g., it's non-obvious to me what "nullTag" means, especially since it seems to be applied to recursively in a way that seems surprising to me—I'm not sure why a list being nullable and its contents being nullable are conceptually a single argument since there are perfectly valid types that can't express.

File: packages/pigeon/lib/java_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:306]: As with the Dart version, a comment would be helpful.

File: packages/pigeon/lib/java_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:328]: Option nit (for here, 331, and 334): I find that using an `else` on an `if` whose body unconditionally returns obfuscates the early return. It's much less obvious with this construction that any code after the if/else if/else chain as written here would never execute (because there's no path that doesn't return).

File: packages/pigeon/lib/objc_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/objc_generator.dart:95]: Comment please.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:415]: s/isn't/aren't/

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:639]: Why is the your new helper class "TypeArgument" rather than a more general "TypeSpecification" (or whatever) so that it can be used for the main type itself? It seems odd to have exactly the construction you need to make these three variables also a single variable, which they conceptually are, but not use it because the class is described as being for the arguments rather than for types in general.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:662]: And then these would be a single parameter to Field.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [reply packages/pigeon/lib/ast.dart]: Done, added the example to the class docstring and reworded this one.

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [reply packages/pigeon/lib/ast.dart:97]: Yea, that's a fair bit more concise but it doesn't match the pattern I'm using elsewhere.  It's just easier and more robust to follow the format and be verbose without thinking about it imo.

File: packages/pigeon/lib/dart_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/dart_generator.dart:188]: At this point you can't have a `List` because that's a primitive type.  Those are blocked by the front end parser.  You can only pass in nonnull custom classes here.  This is probably a relevant issue on the primitive generics PR though.  We can enforce whatever rule we want in the front end parse to match what we support in the generators.  We don't have full nullability support yet, the front end parse will block some usage.

File: packages/pigeon/lib/dart_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/dart_generator.dart:269]: Done, I added a docstring that explains it in prose.

File: packages/pigeon/lib/dart_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/dart_generator.dart:279]: Added a docstring.  I don't know if this is the best place to document `nullTag` it's used all throughout the Dart generator to essentially represent if you are generating nullsafe code or not.  I imagine for people looking at it they might confuse it with actually having some control over nullsafety.  The use cases that are accepted should be asserted in the front end compiler until we have full nullsafety support.  Clear as mud?

File: packages/pigeon/lib/java_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/java_generator.dart:308]: Done.  (I hope to factor this out to generator tools at some point)

File: packages/pigeon/lib/java_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/java_generator.dart:330]: I just played around with the Dart parser to see if I could condense this to one return statement and enforce that the return value is set in each branch but Dart doesn't do that.  I filed a bug here: https://github.com/dart-lang/sdk/issues/46759

If someone added code after the if block they would get a linter error that it is dead code.

File: packages/pigeon/lib/objc_generator.dart
Comment (gaaclarke) [reply packages/pigeon/lib/objc_generator.dart:97]: done

File: packages/pigeon/lib/pigeon_lib.dart
Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:415]: done

File: packages/pigeon/lib/pigeon_lib.dart
Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: I do introduce the concept of the `TypedEntity` in a later PR which allows me to operate on Fields and TypeArguments.  I've been just implementing this as I've needed it.  Once everything is done with the 1.0 features it might be worth revisiting this since there is still some ill fitting stuff like using Fields for return values.  Maybe it means shifting to a more generic concept, or maybe introducing a ReturnType class that implements `TypedEntity`.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:662]: Yep, just mentioned returnType.  Do you want me to file an issue to revisit this?  Otherwise I'll can keep an eye on this.  Check out `TypedEntity`.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/ast.dart:97]: The point of toString is for human consumption; I'm having a hard time imagining a case where I'd want to see the second instead of the first while debugging.

What is the use case where the the second form is easier to read than the first?

Comment (gaaclarke) [reply packages/pigeon/lib/ast.dart:97]: It tells you more in the the same way a debugger does by just telling you the raw data.  A simple sexp parser could parse the output of toString to create a recreation of the data.  Any logic to make it more concise might demonstrate the same bug someone is trying to debug.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:639]: > I've been just implementing this as I've needed it.

But you already do need it in this PR, thus my comment: you have several places where you have exactly the set of things that the new type is meant to encapsulate, and you're now fixing the fact that that they aren't grouped, but only for an arbitrary subset instead of all cases. And then mixing the new (better) way and the old way in the same pieces of code.

Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:639]: I just looked, and `TypedEntity` doesn't actually address my concern here. It doesn't fix the fact that `Field` is storing its type as three separate parameters, it just slaps an interface over it. I'm not suggesting that `Field` should be-a `TypeSpecification`, I'm saying it should have-a `TypeSpecification` instead of three separate members.

Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: I don't actually need it.  I've implemented the feature without it and without duplicating any code.  The reason I factored out the TypedEntity is because I eventually needed to it create logic that would operate over them both so I didn't have to duplicate it.  At this point it's more of a style issue imo.  I think refactoring to a ReturnType would make a lot of sense.  That way I could remove the last of the references to 'void' (I was able to remove those for arguments in a later PR).

Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:639]: > I don't actually need it. I've implemented the feature without it

Obviously I don't mean the code doesn't work without it. I hope we can agree that the purpose of a code review is not to just make the code run though.

> and without duplicating any code.

This is objectively not the case:
`TypeArgument:`
```
class TypeArgument {
  [...]
  /// A string representation of the base datatype.
  final String dataType;

  /// The type arguments to this [TypeArgument].
  final List<TypeArgument>? typeArguments;

  /// True if the type is nullable.
  final bool isNullable;
```

`Field`:
```
class Field extends Node {
  [...]
  /// The data-type of the field (ex 'String' or 'int').
  [...]
  /// True if the datatype is nullable (ex `int?`).
  bool isNullable;
  [...]
  List<TypeArgument>? typeArguments;
```

The code this comment is attached to:
```
    late String argType;
    bool isNullable = false;
    List<TypeArgument>? argTypeArguments;
```

You have at least three different places just in the context of this PR alone where you have a collection of the three elements of the type expressed as three separate variables. One of them is declaring exactly the type that groups them, but artificially restricting its use. That creates both unnecessary duplication, and inconsistency, both of which hurt readability and maintenance.

> At this point it's more of a style issue imo.

I'm curious if you would argue that the difference between these two (slightly pseudo-code) implementations of a path are just a style issue:

```
class Point {
   int x;
   int y;
}

class Path {
  List<Point> points;
}
```

vs.

```
class SubsequentPoint {
  int x;
  int y;
}

class Path {
  int startX;
  int startY;
  List<SubsequentPoint> otherPoints;
}
```

Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: I think we look at things a bit differently.  Code is either defining how data resides in memory or is operating on that data.   Not duplicating yourself is only important for the actual operating logic since that's where maintenance and bugs come from.  The structure definitions are just a means to create hierarchy and organization to help people understand the data and drive the goal of not repeating your logic.  The structures obviously shouldn't try to confuse the data either though.

So you'd like this?
```dart
class TypeDeclaration {
  bool isNullable;
  String name;
  List<TypeDeclaration> typeArguments;
}

class Field {
  TypeDeclaration typeDeclaration;
}
```

I think that's fine.  I can make that change.  In terms of my goal of not duplicating logic, the follow is is just as good.  I do think it's just style.  Neither stops us from sharing logic and neither is confusing.  I would say your suggestion is slightly more clear since it involves less concepts.

```dart
interface TypedEntity {
  bool isNullable;
  String name;
  List<TypeArgument> typeArguments;
}

class TypeArgument implements TypedEntity {}

class Field implements TypedEntity {}
```

Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: (Although, one benefit of the later is that the concepts more closely match the elements in the language, so more of a wash on clarity and we still might need a TypedEntity that just returns a TypeDeclaration)

Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: Okay, I did the refactor, but I realized that it's going to be a headache to rebase the depending PRs.  Mind if I revert this and punt on this until the other things land?  I had to use the refactor tools to manage this safely.

Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:639]: > Not duplicating yourself is only important for the actual operating logic since that's where maintenance and bugs come from. The structure definitions are just a means to create hierarchy and organization to help people understand the data and drive the goal of not repeating your logic.

Bugs come from people, and maintenance is done by people. Making code easier for people to understand makes maintenance easier, and bugs less likely.

> So you'd like this?

Yes. Types are a core object in Pigeon's logic, and having the components of a type clearly, distinctly encapsulated in an object that define a type makes it easier to understand logic that is operating on types.

> I do think it's just style. Neither stops us from sharing logic and neither is confusing.

I find it confusing that a core piece of data for your domain doesn't have a clear object, and that as a result I have to track a triad of variables as being related in multiple sections of code (just as, to go back to my simple example, I would find it confusing and problematic if a codebase in which a point was a core concept had no Point class, and just always used separate x and y variables).

And I find it very confusing that you are expressing the concept of "a thing that has a type" not via a has-a of a Type, which seems like the very obvious mapping, but via interface inheritance of a basket of data.

>  I would say your suggestion is slightly more clear since it involves less concepts.

If even you--as the person most familiar with the code as is because you wrote it all, and thus essentially guaranteed to find your version more readable than anyone else will find it--find my version slightly clearer, and I find my version much clearer, isn't that pretty good evidence that it is, in fact, more readable/understandable?

(There are also other effects. E.g.:
- Your version requires `toString`, conversion to a Dart string, and really any other logic that operates specifically on a type to be either duplicated in each implementation (going against your argument that your version is just as good when it comes to avoiding duplication), or made a freestanding function that takes a `TypedEntity` argument even if it would be clearer as a class method.

  In fact, looking at the next PR again: look closely at [this](https://github.com/flutter/packages/pull/418/files#diff-bda4e51dd560ea4eb7a3ea3dbca764dfb14d3014da1df8955bda4ff864c3603cR112) and [this](https://github.com/flutter/packages/pull/418/files#diff-bda4e51dd560ea4eb7a3ea3dbca764dfb14d3014da1df8955bda4ff864c3603cR147). You have already had to duplicate logic, **and even in that one single line of duplicate logic you already have a bug**.
- Your version does not allow extracting the type of another object like a Field. In fact, you list:
  > we still might need a TypedEntity that just returns a TypeDeclaration
  
  as a downside of my approach, but that's not a concept that's even expressible in your version. You have to always pass around the entire object (by interface) just to operate on the type. And that's not purely stylistic; you couldn't, for instance, have a set of types, or a collection keyed by types.)

> Okay, I did the refactor, but I realized that it's going to be a headache to rebase the depending PRs. Mind if I revert this and punt on this until the other things land?

But the reason it's hard to rebase the other PRs is because you're adding even more code that uses this and more layers around it (like `TypedEntity`). All of the problems I have with the type code in this PR are going to be further compounded in the subsequent PRs, and I wouldn't be able to suggest related improvements in any of that other new code either.

Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: Sorry, just seeing this now.

Well you raise some interesting points.  It seems to be at odds with the age old advice of the "Law of Demeter" which says we should prefer 'field.isNullable` over `field.type.isNullable`.

How about this as a compromise:
```dart
interface TypedEntity {
  bool isNullable;
  String name;
  List<TypeDeclaration> typeArguments;
}

class TypeDeclaration implements TypedEntity {
  bool isNullable;
  String name;
  List<TypeDeclaration> typeArguments;
}

class Field implements TypedEntity {
  TypeDeclaration type;
  bool get isNullable => type.isNullable;
  String get name => type.name;
  List<TypeDeclaration> get typeArguments => type.typeArguments;
}
```

This gives you the structure you want, makes Demeter happy and keeps the interface stable across the backlog of PRs we have.  We can decide if we want to eliminate TypedEntity in the future.

Just to verify, this is the last issue to resolve on this PR, right?  If you don't like this idea let's try to have a chat tomorrow.  I've started working on generating code for multiple arguments since this is at loggerheads.  The sooner we resolve this the less work I 'll have reconciling the existing code.

Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:639]: > Well you raise some interesting points. It seems to be at odds with the age old advice of the "Law of Demeter" which says we should prefer 'field.isNullable`over`field.type.isNullable`.

I've not personally seen someone argue before that the LoD should be so strictly applied that simple data objects should never be used directly, but... 

> How about this as a compromise:

Sure, if you feel strongly about it, wrapping all of the accesses is fine. However you should be using different names for `TypedEntity`'s `name` getter wrapper since it's in the namespace of the enclosing class, not the type object, and is ambiguous in that context. (In particular, you're going to immediately have a bad time in `Field`.)

> Just to verify, this is the last issue to resolve on this PR, right?

As far as I know, yes.

Comment (gaaclarke) [reply packages/pigeon/lib/pigeon_lib.dart:639]: Okay, done.  PTAL.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:103]: I find this description and the examples (and even the name of the getter) misleading. If something is a `List<String>?` its type is that, not `List`. Is there a standard name for this component of the type in Dart? If not, maybe 'baseDataType'?

At the very least, the example should have something like that `List<String>?`->`List` example, rather than only having examples where the two other fields don't apply.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:107]: Why the inconsistency between `type` and `dataType` here? Having two different ways of referring to the same thing in the same API is confusing.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:126]: Same consistency issue here. Although in the context of `TypeDeclaration` it's not clear to me what purpose having a `type` prefix is serving; this could probably just be called `arguments`.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:123]: In this context it should just be called `name` or `baseName` or whatever term you use above; the class is the type.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:114]: Why is this a `TypedEntity`? A `TypedEntity` is, per the above "An entity that represents a typed concept"; i.e., it's a thing that has a type. A type doesn't *have* a type, it *is* a type.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:144]: As above, these should be a `TypeDeclaration`.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:171]: Note that you still have the bug here that will be fixed by replacing the fields here with `TypeDeclaration`.

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:124]: I'm curious: why was avoiding LoD violations so important for `TypeDeclaration` that it needed a new interface wrapping its accessors, while this—which looks conceptually the same to me, and which you do throughout this PR—is not an issue for you?

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:284]: Here and below, consider a local for `field.typeArguments` so that you don't need the `!`. (You could even hoist it outside the switch and share the line; it's not always needed, but it's also just a trivial accessor so calling it when not needed isn't going to hurt anything.)

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/pigeon_lib.dart:658]: I just noticed that this already introduces the same confusing use of `Field` that came up in the next PR's review.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:662]: I called this out as one of the motivations for adding `TypeDeclaration`, so I'm not sure why it isn't using it now.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by gaaclarke** [comment]
@stuartmorgan Okay, how about this.  Let me revert this last change and address your concerns in a PR that sits at the end of all these PRs.  Your criticisms are pervasive and weaves itself through multiple changes.  You don't have to approve this PR until everything is kosher at the end.  How does that sound?   I know it isn't how we typically work but because of the nature of how we are working and the 2 week backlog we have that makes the most sense to me.

**Comment by gaaclarke** [comment]
@stuartmorgan Here is a refactor PR for the AST that I think addresses all your comments: https://github.com/flutter/packages/pull/427  It will be a better use of our time if we punt any discussion about the AST structure until that PR.  That way you get a full picture and and I can easily make changes.

**Comment by stuartmorgan** [comment]
> You don't have to approve this PR until everything is kosher at the end.

If I'm not allow to make any substantive comments on the next 3(? 4?) PRs without constantly context-switching between the review I'm doing and a mega-PR to see if it's addressed there, that's strictly more work, and more cognitive load, than just reviewing one giant PR. So if you're not willing to do the rebasing, then you should just combine them into one PR.

**Comment by gaaclarke** [comment]
subsumed by https://github.com/flutter/packages/pull/428

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [reply packages/pigeon/lib/ast.dart]: That's the nomenclature the dart developers chose for these particular thing.

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [reply packages/pigeon/lib/ast.dart]: That definition doesn't seem contradictory to me.  TypeDeclaration conforms to TypedEntity and the definition of the TypedEntity and its docstring make sense, therefore TypeDeclaration does conform to the docstring of TypedEntity.

**Pull request closed without merging**
